compiler/src/dmd/dsymbolsem.d

```diff
@@ -6812,6 +6830,140 @@ private void aliasAssignSemantic(AliasAssign ds, Scope* sc)
+    {
+        if (otd.building)
+            td = otd;
+        else
+        {
+            td = new TupleDeclaration(tempinst.loc, ttpident, otd.objects.copy());
+            td.building = true;
+            td.storage_class |= STC.templateparameter;
+        }
+    }
+    // If starting from single element in aliassym (td == null) we need to build the tuple
+    // after semanticTiargs to keep same semantics (for example a FuncLiteraldeclaration
+    // template argument is converted to FuncExp)
+    if (td)
+        aliassym.aliassym = td;
+    aliassym.semanticRun = PASS.semanticdone;
+    if (!TemplateInstance.semanticTiargs(tempinst.loc, sc, tempinst.tiargs, 0, td))
+    {
+        tempinst.errors = true;
+        return null;
+    }
+    // The alias will stop tuple 'building' mode when used (in AliasDeclaration.toAlias(),
+    // then TupleDeclaration.getType() will work again)
+    aliassym.semanticRun = PASS.initial;
+
+    if (!td)
+        td = new TupleDeclaration(tempinst.loc, ttpident, tempinst.tiargs);
+    else
+    {
+        auto tiargs = tempinst.tiargs;
+        size_t oldlen = td.objects.length;
+        size_t origstart;
+        size_t insertidx;
+        size_t insertlen;
+        size_t i;
+        foreach (o; *tiargs)
+        {
+            if (o !is td)
+                ++insertlen;
+            else // tuple contains itself (tuple = AliasSeq!(..., tuple, ...))
+            {
+                if (insertlen) // insert any left element before
+                {
+                    td.objects.insert(insertidx, (*tiargs)[i - insertlen .. i]);
+                    if (insertidx == 0) // reset original tuple start point
+                        origstart = insertlen;
+                    insertlen = 0;
+                }
+                if (insertidx) // insert tuple if found more than one time
+                {
+                    td.objects.reserve(oldlen); // reserve first to assert a valid slice
+                    td.objects.pushSlice((*td.objects)[origstart .. origstart + oldlen]);
+                }
+                insertidx = td.objects.length;
+            }
+            ++i;
+        }
+        if (insertlen)
+        {
+            if (insertlen != tiargs.length) // insert any left element
+                td.objects.pushSlice((*tiargs)[i - insertlen .. i]);
+            else
+                // just assign tiargs if tuple = AliasSeq!(nottuple, nottuple...)
+                td.objects = tempinst.tiargs;
+        }
+    }
+    return td;
+}
+
+/***************************************
+ * Check if a template instance is a trivial AliasSeq but without other overloads.
+ * We can only be 100% sure of being AliasSeq after running semanticTiargs()
+ * and findBestMatch() but this optimization must happen before that.
+ */
+private TemplateInstance isAliasSeq(Scope* sc, TypeInstance ti)
+{
+    if(ti.tempinst.findTempDecl(sc, null) && ti.tempinst.tempdecl)
```

Comment:
please use early return

Reply:
Moved out of this function for better error handling.